lib/plugins/aws/plan/index.test.js

```diff
@@ -0,0 +1,142 @@
+'use strict';
+
+const AwsProvider = require('../provider/awsProvider');
+const AwsPlan = require('./index');
+const Serverless = require('../../../Serverless');
+const CLI = require('../../../classes/CLI');
+const expect = require('chai').expect;
+const sinon = require('sinon');
+
+describe('AwsPlan', () => {
+  let awsPlan;
+  let serverless;
+  let options;
+
+  beforeEach(() => {
+    serverless = new Serverless();
+    options = {
+      stage: 'dev',
+      region: 'us-east-1',
+    };
+    serverless.setProvider('aws', new AwsProvider(serverless, options));
+    serverless.config.servicePath = 'foo';
+    serverless.cli = new CLI(serverless);
+    awsPlan = new AwsPlan(serverless, options);
+  });
+
+  describe('#constructor()', () => {
+    it('should set the serverless instance', () => expect(awsPlan.serverless).to.equal(serverless));
+
+    it('should set options', () => expect(awsPlan.options).to.equal(options));
+
+    it('should set the provider variable to an instance of AwsProvider', () =>
+      expect(awsPlan.provider).to.be.instanceof(AwsProvider));
+
+    it('should have commands', () => expect(awsPlan.commands).to.be.not.empty);
+
+    it('should have hooks', () => expect(awsPlan.hooks).to.be.not.empty);
```

Comment:
Currently we're trying to not introduce such tests for new functionality (sorry that it's not properly communicated at this point).

Problem with tests like that, is that they do not really test outcome (so what's important) but characteristics of internal implementation, which should rather be treated as a black box, and such tests make eventual future refactor really painful.

So instead of that, we now use [`runServerless`](https://github.com/serverless/test/blob/master/docs/run-serverless.md) util, which initialize complete `serverless` instance (against pre-configured service, which can be passed either via inline config or pre-prepared fixture), and test the outcome of that.

Good examples to check:
- https://github.com/serverless/serverless/blob/80f1c798a293d99529f9e7200227aa0f96ace219/lib/plugins/aws/package/compile/events/httpApi/index.test.js
- https://github.com/serverless/serverless/blob/80f1c798a293d99529f9e7200227aa0f96ace219/lib/plugins/aws/package/compile/functions/index.test.js#L2624-L2718

As when testing `plan` you'd need to mock AWS requests, see how it's done here: https://github.com/serverless/serverless/blob/656c0c5abc3748b10efbacc753ea536bc8ca773a/lib/plugins/aws/deploy/lib/uploadArtifacts.test.js#L18-L39

Ideally for plan, we should have one test file `plan/index.test.js` (no need to create test files for all lib files), in which we should test most common _plan_ scenarios (no need to not dive too much into corner cases), by confirming on generated output which as I can can be read from `plugin.analysis`.

If you have any doubts, questions, please ask. I don't want this to feel as blocker

Reply:
Many thanks for the recommendation!, it's a very useful approach!. I refactored the test of the entry point to use a slightly different version of this approach. I don't see the mechanism to mock aws, as a result I implemented a very similar mechanism that allows to use sinon to mock aws.
If you are OK I'll keep the unit tests and I'll refactor them to this kind of functional test as changes are needed or any bug is identified. 
Please let me know if this is OK or if I should remove the unit tests.
proposed fix is committed in https://github.com/serverless/serverless/pull/7509/commits/6e05e24be27735e43af1aab19fea39798e0ad51a